src/controllers/http/worker-http-controller.ts

```diff
@@ -0,0 +1,191 @@
+
+import { Worker } from "bullmq";
+import { Redis, Cluster } from "ioredis";
+import { debug } from "../../utils/log";
+
+import { HttpHandlerOpts, WorkerMetadata } from "../../interfaces";
+import { validateWorkerMetadata } from "../../validators";
+import { config } from "../../config";
+
+const debugEnabled = config.debugEnabled;
+
+const workers: { [queueName: string]: Worker } = {};
+
+const workerMetadataKey = config.workerMetadataKey;
+
+// Gracefully close all workers
+process.on('exit', async () => {
+  for (const queueName in workers) {
+    await workers[queueName].close();
```

Comment:
would that work fine if all workers share same connection?

Reply:
Yes. The shared connection will not be closed, but all workers also have a dedicated connection for blocking calls, this dedicated connection needs to be closed by calling worker.close().

Reply:
How come? `worker.close()` does this:
```
                .finally(() => client.disconnect())
                .finally(() => this.connection.close())
```
`client` here seems to be the shared connection.

(that's actually something that I addressed in https://github.com/taskforcesh/bullmq/pull/2449, having encountered exactly this in our app, workers closing shared connection)

Reply:
But client in this context is actually a duplication of the client that you pass to the constructor, so it is safe to close it...
```ts
      const client =
        this.blockingConnection.status == 'ready'
          ? await this.blockingConnection.client
          : null;
```

Reply:
```ts
this.blockingConnection = new RedisConnection(
      isRedisInstance(opts.connection)
        ? (<Redis>opts.connection).duplicate({ connectionName })
        : { ...opts.connection, connectionName },
      false,
      true,
      opts.skipVersionCheck,
    );
```

Reply:
Oh, I see, so it effectively always creates a new connection for each worker.

Is there a reason for that? There are limitations for shared connections that wouldn't work for BullMQ?

Reply:
Because we need connections that block, and a blocked connection can not be shared as no commands will be send until the connection is unblocked.